### DISCUSSION TITLE
Hidden file scanning support with an option

### DISCUSSION THREAD

**Discussion by soullivaneuh** [squizlabs/PHP_CodeSniffer#3929] [Ideas]
# Summary

Currently, phpcs is completely ignoring hidden file from its scanning process:

```
# Files have the same content.
❯ ls -la .test
-rw-rw-r-- 1 sullivan sullivan   75 nov.  24 12:17 .test.php
-rw-rw-r-- 1 sullivan sullivan   75 nov.  24 12:17 test.php
# Only test.php is scanned
❯ php phpcs.phar .test

FILE: /home/sullivan/p/gitlab.com/ss-open/ci/recipes/.test/test.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 2 | ERROR | Missing file doc comment
----------------------------------------------------------------------

Time: 65ms; Memory: 6MB
# The hidden file is ignored even when it is directly specified
❯ php phpcs.phar .test/.test.php 
```

By digging on the code, I figured out it looks to be intended: https://github.com/squizlabs/PHP_CodeSniffer/blob/7763e2e1f773cb0615ed8afa133189fc804f583d/src/Ruleset.php#L634-L637

Also, this problematic was already reported in #1928 but closed without solution because it will introduce a breaking change that not worth it regarding the interest of the feature.

## Concern and interest

In my opinion, considering the scan of hidden files "because it is worthless" is a matter of opinion and we should let the user decide about that.

For my case, I have some special hidden files that are written in PHP and I would like them to be scanned by the linter as well.
This is, for example, the case of the php-cs-fixer's configuration file, named `.php-cs-fixer.dist.php`.

@gsherwood On #1928, you said this kind of feature addition will introduce a breaking change. May you please elaborate about which kind of breaking change would be introduced here? I am assuming the breaking change would be the tool being scanning additional files "by itself" once updated. Am I right?

## Feature proposal

The proposal is simple: Let the tool being able to scan hidden files. According to what I found above, the implementation would be quite simple, meaning adapting the skip condition.

However, we still have the breaking change concern. For that, I propose to add CLI options named `--include-hidden-files` and `--skip-hidden-files` (or similar).

For the current major version (v3), the `--skip-hidden-files` would be the default option, which will produce the same behavior as before.

On the next major version (v4), I suggest to make `--include-hidden-files` the default which make more sense to me, regarding my statement about the "matter of opinion" fact.

What do you think about that?

Thanks for the reading!

**Comment by jrfnl** [comment]
Loosely related to #2916.

I can imagine a change that _if_ - and only _if_ - a dot-prefixed file is explicitly passed using the `<file>` syntax, the file would be scanned as of v4. Adding the CLI options is IMO not the way to go.
